src/sidebar/build-thread.js

```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
  * the current visibility filters and sort function and returns
  * the thread structure that should be rendered.
  *
+ * An Annotation present in `annotations` will not be present in the returned threads if:
+ * - The annotation does not match thread-level filters (options.threadFilterFn), OR
+ * - The annotation is not in the current selection (options.selected), OR
+ * - The annotation's thread is hidden and has no visible children
+ *
+ * Annotations that do not match the currently-applied annotation filters
+ * (options.filterFn) will have their thread's `visible` property set to `hidden`
+ * (an exception is made if that annotation's thead has been forced visible by
+ * a user).
+ *
  * @param {Array<Annotation>} annotations - A list of annotations and replies
- * @param {Options} opts
+ * @param {Partial<Options>} options
  * @return {Thread} - The root thread, whose children are the top-level
  *                    annotations to display.
  */
-export default function buildThread(annotations, opts) {
-  opts = Object.assign({}, defaultOpts, opts);
+export default function buildThread(annotations, options) {
+  const opts = { ...defaultOpts, ...options };
+
+  const annotationsFiltered = !!opts.filterFn;
+  const threadsFiltered = !!opts.threadFilterFn;
```

Comment:
The `threadsFiltered`, `hasSelection` and `hasHighlights` variables are used only once so I think it would be a marginally simpler just to replace the associated `if (...)` conditions with their initializer, minus the `!!` to-boolean coercion.

Reply:
I kind of like this the way it is as it is an explicit set of factors that affect state all in one place. It makes the _intent_ of certain logic easier to understand, IMHO. For example, a later expression:

        const hasUnfilteredChildren =
          annotationsFiltered && hasVisibleChildren(thread);

It's (marginally :) ) easier to understand the intent here vs.

        const hasUnfilteredChildren =
          opts.filterFn && hasVisibleChildren(thread);

I know that `annotationsFiltered` wasn't one of the references you called out here, but I like the completeness of being explicit about what it means for threads to be filtered, if we're defining what it means for annotations to be filtered.

The code I started out with also wasn't clear about delineating between annotation filtering and thread filtering, which I think this helps define.